### ISSUE TITLE
Parse could return Result

### ISSUE THREAD

**Issue by CrockAgile** [snewt/bnf#12]
Parsing a grammar can fail. Currently, if parsing fails an empty grammar is returned and errors are printed to stdout. Users would likely be interested in detecting if the parsing failed, but capturing `stdout` or `stderr` is not ideal. The code required to currently detect an empty grammar would look something like:

```rust
let grammar = bnf::parse(input);
// errors printed to stdout
if grammar.productions.is_empty() {
    println!("empty grammar");
} else {
    println!("non-empty grammar");
}
```

Instead of printing error strings to stdout or stderr, they could be included in the `Result<Grammar,Error>` type, allowing for more traditional error handling:

```rust
let grammarResult = bnf::parse(input);

match grammarResult {
    Ok(grammar) => println!("working grammar: {}", grammar),
    Err(e) => eprintln!("grammar parsing failed: {}", e),
}
```

If this change were to be made, would reporters still be involved? As a library it seems strange to output to stdout or stderr instead of output being handled in the code. But as a command line tool, error reporting in terminal is definitely valuable. Maybe BNF should include an optional CLI binary intended for terminal use? I forget if cargo allows for dual binary & library crates.

**Comment by snewt** [comment]
Good point, I agree that relying on `stdout` or `stderr` isn't totally reasonable. Probably providing the error report in the `bnf::parse` would be the cleaner and more intuitive functionality for anyone incorporating the library. Thanks for raising the issue, this one seems like it should have a reasonably high priority but I'm slightly concerned with introducing a breaking change just 3 days after first publishing the crate haha :cold_sweat:.

**Comment by CrockAgile** [comment]
Closed by PR #24
